crates/task-impls/src/consensus.rs

```diff
@@ -1220,6 +1220,7 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
             let block_header = TYPES::BlockHeader::new(
                 commit_and_metadata.commitment,
                 commit_and_metadata.metadata.clone(),
+                &consensus.instance_state,
```

Comment:
At some point we may want to collect all these arguments into a struct, since for various implementations a lot of them might be unused anyways. Plus, I believe the arguments to `validate_and_apply_header` should be the same as the arguments to `Header::new`, plus the proposed header?

Reply:
Looks like the only arguments this function and `validate_and_apply_header` have in common are the instance state and parent header. I'm not sure if adding a struct with two fields would be useful. Or do you mean a struct with more fields not necessarily used by `validate_and_apply_header`?

Reply:
Hmm you're right. In *principle*, `validate_and_apply_header` and `Header::new` should get the same exact inputs, because the former needs to check that whatever the latter did was done correctly. But, for `validate_and_apply_header`,
* `state` is given as `self`
* `commitment` and `metadata` are part of `proposed_header`
So we are left with only `instance_state` and `parent_header` explicitly in common.

_Maybe_ to make this analogy more explicit we could make the order of arguments more consistent, e.g.
* Make `state` the first argument to `BlockHeader::new`, since `self` is first in `validate_and_apply_header` (alternative idea: make `new_header` a method of `State`)
* Make `instance` the second argument to both
* Move `parent_header` before `proposed_header` since `parent_header` is an argument to both but `proposed_header` only applies to `validate_and_apply_header`
* Delete `view_number` from `validate_and_apply_header` since we don't get it here, we don't really need it, and it is probably best for it to remain internal to consensus

Anyway, this is just some logic that I am using to check that we have all the necessary arguments in both places, I guess it's not really necessary to make this logic explicit in the code but I think it would be nice to have.

Reply:
Makes sense! I think it's nice to have a more consistent and reasonable ordering of arguments. I'm also going to make `commitment` and `metadata` the last arguments to `BlockHeader::new`.